compose/ui/ui/src/skikoMain/kotlin/androidx/compose/ui/ComposeScene.skiko.kt

```diff
@@ -429,26 +423,49 @@ class ComposeScene internal constructor(
     }
 
     private fun onMousePressed(event: PointerInputEvent) {
-        if (focusedOwner.isAbove(hoveredOwner)) {
+        val owner = hoveredOwner(event)
+        if (focusedOwner.isAbove(owner)) {
             focusedOwner?.onDismissRequest?.invoke()
         } else {
-            hoveredOwner?.processPointerInput(event)
+            owner?.processPointerInput(event)
             mousePressOwner = focusedOwner
         }
     }
 
     private fun onMouseReleased(event: PointerInputEvent) {
-        val owner = mousePressOwner ?: hoveredOwner
+        val owner = mousePressOwner ?: hoveredOwner(event)
         owner?.processPointerInput(event)
     }
 
-    private fun onMouseScrolled(event: PointerInputEvent) {
-        if (!focusedOwner.isAbove(hoveredOwner)) {
-            hoveredOwner?.processPointerInput(event)
+    private fun onMouseMove(event: PointerInputEvent) {
+        val owner = if (event.buttons.areAnyPressed) mousePressOwner else hoveredOwner(event)
+
+        if (owner != null && mouseEnterExitOwner != null && owner != mouseEnterExitOwner) {
+            mouseEnterExitOwner?.processPointerInput(event.copy(eventType = PointerEventType.Exit))
```

Comment:
Looks somewhat clumsy. Can we make it more transparent?

Reply:
What you mean?

Reply:
Well, it's not too what this code does and why. Also too many things happen on high frequency mouse move event.

Reply:
>  Also too many things happen on high frequency mouse move event.

We can't get rid of this logic, because we should hit test every move event.

Also, before do any optimizations, we should measure.

The new code in this PR shouldn't make things worse (in terms of performance), because the new code only called one time - when we move from one owner to another.

> Well, it's not too what this code does and why

Any suggestions, how can we make it better? In my opinion this code is pretty straightforward - we force exit/enter when the owner behind the cursor changes.

As for the other logic in this class for events, I agree, we can make the code simpler, if we think thoroughly about all cases, but that is not simple, and out of scope of this PR.

Reply:
Refactored the code. Now all enter/exit event are treated as Move events. Enter/Exit we determine ourselves, by owner's bounds